### PR TITLE
[WIP] Data source dependencies

### PR DESCRIPTION
**Pull request by SwampDragons**
Allows data sources to use other datasources. 

I implemented this with something hopefully a little more elegant than a brute force "retry" solution. Instead I check the attributes on the data source to see whether it needs other data sources; if it does, I wait to evaluate this data source until after all the others have been launched and executed. Then, I recursively evaluate the data sources that remain, with a recursion depth of 10 to prevent getting stuck in infinite cycles. 

This PR also adds a data source for testing, the "null" data source.

It also adds tests. 

TODO: Integration tests using the "null" data source; I have one written locally but want to add a couple into the testacc suite. I'll be doing this in a separate PR since this one is already pretty big. The test file is 

```
source "null" "example" {
  communicator = "none"
}

data "null" "foo" {
  input = "chocolate"
}

data "null" "bar" {
  input = "vanilla"
}

data "null" "baz" {
  input = "${data.null.foo.output}-${data.null.bar.output}-swirl"
}

data "null" "bang" {
  input = "${data.null.baz.output}-with-marshmallows"
}
locals {
  flavor           = "mint"
  ice_cream_flavor = "${local.flavor}-ice-cream"
  foo              = "${data.null.foo.output}"
  bar              = "${data.null.bar.output}"
  baz              = "${data.null.baz.output}"
  bang              = "${data.null.bang.output}"
}

build {
  name = "mybuild"
  sources = [
    "source.null.example"
  ]
  provisioner "shell-local" {
    inline = [
      "echo data is ${data.null.foo.output}",
      "echo data is ${data.null.bar.output}",
      "echo data is ${data.null.baz.output}",
      "echo data is ${data.null.bang.output}"
    ]
  }
}

```

### CODE REVIEW

File: hcl2template/testdata/datasources/recursive.pkr.hcl [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+data "null" "bar" {
+  input = "vanilla"
+}
+
+data "null" "baz" {
+  input = "${data.null.foo.output}-${data.null.bar.output}-swirl"
+}
+
```
Comment (sylviamoss) [inline hcl2template/testdata/datasources/recursive.pkr.hcl:12]: What about adding one that it's in the wrong order? 
```suggestion
data "null" "yummy" {
  input = "${data.null.bang.output}-and-cinamon"
}
```

Comment (SwampDragons) [reply hcl2template/testdata/datasources/recursive.pkr.hcl:12]: Nice idea!

Comment (SwampDragons) [reply hcl2template/testdata/datasources/recursive.pkr.hcl:12]: I should also do one that actually triggers the max recursion depth, and one that reads data sources from other files.

**Review by sylviamoss — COMMENTED** [review]
This is looking very good! Thanks for the comments, it helped a lot during review 👍🏼

File: hcl2template/testdata/datasources/recursive.pkr.hcl (hunk 1 above)
Comment (sylviamoss) [reply hcl2template/testdata/datasources/recursive.pkr.hcl:12]: Oh yeah, that will be nice!

File: datasource/hcp-packer-image/data.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+						BuildID:       build.ID,
+						IterationID:   build.IterationID,
+						PackerRunUUID: build.PackerRunUUID,
+						ID:            image.ID,
+						ImageID:       image.ImageID,
+						Region:        image.Region,
+					}
+				}
+			}
+		}
```
Comment (nywilken) [inline datasource/hcp-packer-image/data.go:149]: nit: Use guard  clause over the nested if statements.
```go
	if build.CloudProvider != d.config.CloudProvider {
		continue
	}
	for _, image := range build.Images {
		if image.Region != d.config.Region {
			continue
		}
		// This is the desired image.
		output = DatasourceOutput{
			CloudProvider: build.CloudProvider,
			ComponentType: build.ComponentType,
			CreatedAt:     image.CreatedAt.String(),
			BuildID:       build.ID,
			IterationID:   build.IterationID,
			PackerRunUUID: build.PackerRunUUID,
			ID:            image.ID,
			ImageID:       image.ImageID,
			Region:        image.Region,
		}
	}
```

File: hcl2template/testdata/datasources/recursive.pkr.hcl (hunk 1 above)
Comment (nywilken) [reply hcl2template/testdata/datasources/recursive.pkr.hcl:12]: I believe this is already part of the datasouce agreement, but I would add a test for a set of data sources that depend on each other. This should be an error right? 

```
data "null" "gummy" {
  input = "${data.null.bear.output}"
}
data "null" "bear" {
  input = "${data.null.gummy.output}"
}
```

**Review by nywilken — COMMENTED** [review]
This is off to a good start. A couple of suggestions, and request for a new test case.

File: hcl2template/testdata/datasources/recursive.pkr.hcl (hunk 1 above)
Comment (SwampDragons) [reply hcl2template/testdata/datasources/recursive.pkr.hcl:12]: Great call -- I think that'll trigger the "max recursion depth" I referenced above, so it'll knock out both cases.

Comment (SwampDragons) [reply hcl2template/testdata/datasources/recursive.pkr.hcl:16]: Done!

File: datasource/hcp-packer-image/data.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	var errs *packersdk.MultiError
+
+	if d.config.Bucket == "" {
+		errs = packersdk.MultiErrorAppend(errs, fmt.Errorf("The `bucket_name` must be specified"))
+	}
+	if d.config.IterationID == "" {
+		errs = packersdk.MultiErrorAppend(errs, fmt.Errorf("The `iteration_id`"+
+			" must be specified. If you do not know your iteration_id, you "+
+			"can retrieve it using the bucket name and desired channel using"+
+			" the hcp-packer-iteration provider"))
```
[the remark this answers is not in the document]
Comment (SwampDragons) [reply datasource/hcp-packer-image/data.go:59]: Good catch!

File: datasource/hcp-packer-image/data.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	"log"
+
+	"github.com/zclconf/go-cty/cty"
+
+	"github.com/hashicorp/hcl/v2/hcldec"
+	"github.com/hashicorp/packer-plugin-sdk/common"
+	"github.com/hashicorp/packer-plugin-sdk/hcl2helper"
+	packersdk "github.com/hashicorp/packer-plugin-sdk/packer"
+	"github.com/hashicorp/packer-plugin-sdk/template/config"
+	packerregistry "github.com/hashicorp/packer/internal/packer_registry"
```
[the remark this answers is not in the document]
Comment (SwampDragons) [reply datasource/hcp-packer-image/data.go:17]: Ok! I'll resolve conflicts once that's merged.

File: datasource/hcp-packer-image/data.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	// first created, and is unique to this iteration.
+	IterationID string `mapstructure:"iteration_id"`
+	// The UUID associated with the Packer run that created this image.
+	PackerRunUUID string `mapstructure:"packer_run_uuid"`
+	// The iteration id. This is a ULID, which is a unique identifier similar
+	// to a UUID. It is created by the HCP Packer Registry when an iteration is
+	// first created, and is unique to this iteration.
+	ID string `mapstructure:"id"`
+	// ID or URL of the remote cloud image as given by a build.
+	ImageID string `mapstructure:"image_id"`
```
[the remark this answers is not in the document]
Comment (SwampDragons) [reply datasource/hcp-packer-image/data.go:102]: Good point.

File: datasource/hcp-packer-image/data.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+	// The iteration id. This is a ULID, which is a unique identifier similar
+	// to a UUID. It is created by the HCP Packer Registry when an iteration is
+	// first created, and is unique to this iteration.
+	IterationID string `mapstructure:"iteration_id"`
+	// The UUID associated with the Packer run that created this image.
+	PackerRunUUID string `mapstructure:"packer_run_uuid"`
+	// The iteration id. This is a ULID, which is a unique identifier similar
+	// to a UUID. It is created by the HCP Packer Registry when an iteration is
+	// first created, and is unique to this iteration.
+	ID string `mapstructure:"id"`
```
[the remark this answers is not in the document]
Comment (SwampDragons) [reply datasource/hcp-packer-image/data.go:100]: That's fair. We can do that.

File: datasource/hcp-packer-image/data.go (hunk 2 above)
Comment (SwampDragons) [reply datasource/hcp-packer-image/data.go:149]: Sure!

**Comment by SwampDragons** [comment]
messed up my branches -- see #11248 for new PR.
